src/runtime/server.ts

```diff
@@ -1,7 +1,80 @@
-import { Runtime } from "./runtime.ts";
+import { RuntimeError } from "./mod.ts";
+import { HttpData, Runtime } from "./runtime.ts";
 
 const MODULE_CALL = /^\/modules\/(?<module>\w+)\/scripts\/(?<script>\w+)\/call\/?$/;
 
+function notFoundResponse(): Response {
+	return new Response(
+		JSON.stringify({
+			"message": "Route not found. Make sure the URL and method are correct.",
+		}),
+		{
+			headers: {
+				"Content-Type": "application/json",
+			},
+			status: 404,
+		},
+	);
+}
+
+function methodNotAllowedResponse(): Response {
+	return new Response(
+		JSON.stringify({
+			"message": "Method not allowed.",
+		}),
+		{
+			headers: {
+				"Content-Type": "application/json",
+			},
+			status: 405,
+		},
+	);
+}
+
+async function tryAllTypes(req: Request, dataTypes: HttpData[]): Promise<unknown> {
+	const expected: string[] = [];
+	for (const dataType of dataTypes) {
+		try {
+			return await tryGetData(req, dataType);
+		} catch {
+			switch (dataType) {
+				case "json":
+					expected.push("a valid JSON body");
+					break;
+				case "form":
+					expected.push("a valid form body");
+					break;
+				case "query":
+					expected.push("valid query parameters");
+					break;
+			}
+		}
+	}
+
+	const output = {
+		message: `Request must have ${expected.join(" or ")}`,
+	};
+
+	return new Response(JSON.stringify(output), {
+		status: 400,
+		headers: {
+			"Content-Type": "application/json",
+			"Access-Control-Allow-Origin": "*",
+		},
+	});
+}
+
+async function tryGetData(req: Request, dataType: HttpData): Promise<unknown> {
+	switch (dataType) {
+		case "json":
+			return await req.clone().json();
```

Comment:
Is there a reason we need to clone the request here? I could imagine this being expensive for requests with large payloads

Reply:
The reason is that calling `.json()` or `.formData()` on requests consumes the body of the request, which I don't know if there's a simple way to cache and still call things like `.formData()` on otherwise.

Probably needs to have a better solution though.